astropy/wcs/utils.py

```diff
@@ -118,6 +140,60 @@ def _wcs_to_celestial_frame_builtin(wcs):
                 representation_type=SphericalRepresentation,
                 obstime=wcs.wcs.dateobs or None,
             )
+        elif (
+            "LN" in xcoord
+            and "LT" in ycoord
+            and "H" not in xcoord
+            and "CR" not in xcoord
+        ):
+            # Coordinates on a planetary body, as defined in
+            # https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2018EA000388
+
+            object_name = SOLAR_SYSTEM_OBJ_DICT.get(xcoord[:2])
+            if object_name is None:
+                raise KeyError(f"unknown solar system object abbreviation {xcoord[:2]}")
+
+            a_radius = wcs.wcs.aux.a_radius
+            b_radius = wcs.wcs.aux.b_radius
+            c_radius = wcs.wcs.aux.c_radius
+            if "bodycentric" in wcs.wcs.name.lower():
+                baserepresentation = BaseBodycentricRepresentation
+                representation_type_name = "BodycentricRepresentation"
+            else:
+                baserepresentation = BaseGeodeticRepresentation
+                representation_type_name = "GeodeticRepresentation"
+
+            if a_radius == b_radius:
+                equatorial_radius = a_radius * u.m
+                flattening = (a_radius - c_radius) / a_radius
+            else:
+                raise NotImplementedError(
+                    "triaxial systems are not supported at this time."
+                )
+
+            # create a new representation class
+            representation_type = solar_system_body_representation_type.get(xcoord[:2])
```

Comment:
If we keep a cache (which I think we should), we have to be sure it is correct. So, we need a new instance any time a new object *or* a different representation *or* new equatorial/radius are used. That means we should set the key to `object_name, baserepresentation, equatorial_radius, flattening`.

Thinking more about this, I think I would implement it with `lru_cache` from [functools](https://docs.python.org/3/library/functools.html), as follows:
```
from functools import lru_cache

@lru_cache(maxsize=100)
def solar_system_body_representation_type(object_name, baserepresentation, equatorial_radius, flattening):
    return type(
        f"{object_name}{baserepresentation.__name__[4:]},
        (baserepresentations,),
        dict(_equatorial_radius=equatorial_radius, _flattening=flattening)
    )
```